Add GPU implementation of SparseReshape

benbarsdell opened the pull request:
This follows https://github.com/tensorflow/tensorflow/pull/46275.

cc @nluehr

benbarsdell committed:
Add GPU implementation of SparseReshape

sanjoy commented on tensorflow/python/kernel_tests/sparse_reshape_op_test.py:
```diff
@@ -94,7 +94,7 @@ def testPropagatesFullyKnownDenseShapeWhenShapePartiallyKnown(self):
 
   def testSameShape(self):
-    with self.session(use_gpu=False) as sess:
+    with self.session(use_gpu=True) as sess:
```
You no longer need to set `use_gpu=True` explicitly, it default to True.

sanjoy commented on tensorflow/core/kernels/reshape_util_gpu.cu.cc:
```diff
@@ -0,0 +1,110 @@
+    const Tindex* __restrict__ output_shape,
+    const Tindex* __restrict__ input_indices,
+    Tindex* __restrict__ output_indices) {
+  GPU_1D_KERNEL_LOOP(sparse_index, nnz) {
+    const Tindex* input_index = &input_indices[sparse_index * input_rank];
+    Tindex* output_index = &output_indices[sparse_index * output_rank];
+    Tindex dense_index = 0;
+    // Flatten input index from slowest- to fastest-changing dimension.
+    for (int i = 0; i < input_rank; ++i) {
+      dense_index = dense_index * input_shape[i] + input_index[i];
+    }
+    // Compute output index from fastest- to slowest-changing dimension.
+    for (int i = output_rank; i-- > 0;) {
+      Tindex output_size = output_shape[i];
+      output_index[i] = dense_index % output_size;
+      dense_index /= output_size;
+    }
```
Do you need to care about integer overflow in case the indices are in int32?  Maybe always use int64 for `dense_index`?

sanjoy commented on tensorflow/core/kernels/reshape_util_gpu.cu.cc:
```diff
@@ -0,0 +1,110 @@
+      dense_index = dense_index * input_shape[i] + input_index[i];
+    }
+    // Compute output index from fastest- to slowest-changing dimension.
+    for (int i = output_rank; i-- > 0;) {
```
Please use the more idiomatic `i >= 0; i--`. :)

sanjoy commented on tensorflow/core/kernels/reshape_util_gpu.cu.cc:
```diff
@@ -0,0 +1,110 @@
+  auto config = GetGpuLaunchConfig(nnz, device);
+  return GpuLaunchKernel(ReshapeSparseTensorKernel<int64>, config.block_count,
+                         config.thread_per_block, 0, device.stream(), nnz,
+                         /* input_rank = */ input_rank,
```
Please spell these as `/*input_rank=*/`; our internal tooling will then check that the param names match up.

benbarsdell committed:
Address minor review comments for PR 47251

- Remove use_gpu=True because it is already the default.
- Use int64 for dense_index inside kernel to avoid integer overflow.
- Change reverse for-loop style.
- Reformat inline comments to ensure internal tooling picks them up.
